docs/cvl/methods.md

```diff
@@ -57,36 +65,131 @@ method_summary   ::= "ALWAYS" "(" value ")"
+
+Each entry in the methods block contains a pattern that matches some set of
+contract functions.
+
+ - {ref}`exact-methods-entries` match a single method of a single contract.
+ - {ref}`wildcard-methods-entries` match a single method signature on all contracts.
+
+(exact-methods-entries)=
+### Exact entries
+
+An exact methods block entry matches a single method of a single contract.
+If the contract name is omitted, the default is `currentContract`.
+For example,
+```cvl
+methods {
+    function C.f(uint x) external returns(uint);
+}
+```
+will match the external function `f` of the contract `C`.
+
+Exact methods block entries must include a return type; the Prover will check
+that the declared return type matches the return type of the contract function.
+
+Exact entries may contain {ref}`summaries <summaries>`, {ref}`envfree`, and
+{ref}`optional`.
+
+(wildcard-methods-entries)=
+### Wildcard entries
 
+```{versionadded} 4.0
+Wildcard entries were {ref}`introduced with CVL 2 <cvl2-wildcards>`.
 ```
 
-See {doc}`types` for the `evm_type` and `cvl_type` productions.  See {doc}`basics`
-for the `id` production.  See {doc}`statements` for the `block` production, and
-{doc}`expr` for the `expression` production.
+A wildcard entry matches any function in any contract with the indicated name,
+argument types, and visibility.
+For example,
+```cvl
+methods {
+    function _.f(uint x) external => NONDET;
+}
+```
+will match any external function called `f(uint)` in any contract.
+
+Wildcard entries must not declare a return type, since different matched
+methods may return different types.
+
+Wildcard entries may not have {ref}`envfree` or {ref}`optional`; their only
+purpose is {ref}`summarization <summaries>`.  Therefore, wildcard entries must
+have a summary.
+
+### Location annotations
+
+```{versionadded} 4.0
+Location annotations were {ref}`introduced with CVL 2 <cvl2-locations>`.
+```
+
+Methods block entries for internal functions must contain either `calldata`,
+`memory`, or `storage` annotations for all arguments with reference types (such
+as arrays).
+
+For methods block entries of external functions the location annotation must be
+omitted unless it's the `storage` annotation on an external library function, in
+which case it is required (the reasoning here is to have the information required
+in order to correctly calculate a function's sighash).
+
+(methods-visibility)=
+### Visibility modifiers
+
+```{versionadded} 4.0
+Visibility modifiers were {ref}`introduced with CVL 2 <cvl2-visibility>`.
+```
+
+Entries in the methods block must be marked either `internal` or `external`; the
+entry will only match a function with the indicated visibility.
+
+If a function is declared `public` in Solidity, then the Solidity compiler
+creates an internal implementation method, and an external wrapper method that
+calls the internal implementation.  Therefore, you can summarize a `public`
+method by marking the summarization `internal`.
```

Comment:
Maybe connect with the next paragraph? The flow confused me a bit (maybe because I have the context)

Reply:
See if the revision is better?